### PR TITLE
Dotnet tool for script generation

### PR DESCRIPTION
**Pull request by HEskandari**
- New project for cmdline tool to generate scripts
- Refactor to remove the special logging step (only used for the 'saga generation' step)

### CODE REVIEW

File: src/CommandLine/NServiceBus.Persistence.Sql.CommandLine.csproj [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    <OutputType>Exe</OutputType>
+    <ToolCommandName>sql-persistence</ToolCommandName>
+    <PackAsTool>True</PackAsTool>
+    <Description>.NET Core global tool to generate scripts for NServiceBus Sql persistence</Description>
+    <GenerateDocumentationFile>false</GenerateDocumentationFile>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="McMaster.Extensions.CommandLineUtils" Version="3.1.0" />
+    <PackageReference Include="Particular.Packaging" Version="0.9.0" PrivateAssets="All" />
```
Comment (mikeminutillo) [inline src/CommandLine/NServiceBus.Persistence.Sql.CommandLine.csproj:14]: We might be getting this from the `Directory.Build.props` anyway, in which case we shouldn't have it here.

File: src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs [hunk 2]
```diff
@@ -5,4 +5,4 @@
     MySqlScripts = true,
     PostgreSqlScripts = true,
     OracleScripts = true,
-    ScriptPromotionPath = "$(SolutionDir)$(ProjectDir)Postfix")]
\ No newline at end of file
+    ScriptPromotionPath = @"C:\Temp\Prod")]
```
Comment (mikeminutillo) [inline src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs:8]: I wonder if we should allow the user to specify arbitary variables and then have the msbuild script pass in these two. i.e. if we could say at the command line:

```
sql-persistence general assemblyname outputpath -v SolutionDir="Path\To\Solution" -v ProjectDir="Path\To\Project"
```

**Review by mikeminutillo — COMMENTED** [review]
So far so good. I think the biggest thing is going to be deciding how and where this gets configured. Where precisely is the line between build-time configuration (msbuild) and runtime configuration (command line)? Can we override the attribute baked into the assembly if there's a good reason to do so? Will we deprecate the attribute and allow these settings to be controlled directly via .csproj properties?

File: src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs (hunk 2 above)
Comment (HEskandari) [reply src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs:8]: Problem with this is that, if the tool is used outside the dev environment there may be no Solution/Project directory. It hinders the usage outside of the MSBuild integration.

Alternatively, the relative paths can be used, which means it'll be relative to the project directory (I think).

File: src/CommandLine/NServiceBus.Persistence.Sql.CommandLine.csproj (hunk 1 above)
Comment (HEskandari) [reply src/CommandLine/NServiceBus.Persistence.Sql.CommandLine.csproj]: removed

File: src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs (hunk 2 above)
Comment (HEskandari) [reply src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs]: Since we decided to keep the cmd tool separate to the MSBuild task, there'll be no need for this for now.

**Comment by HEskandari** [comment]
@mikeminutillo had a bunch of commits to fix some broken / non-deterministic tests. I have added a couple of more options. Would be great if you could have another look.

File: src/ScriptBuilder/Writers/OutboxWriter.cs [hunk 3]
```diff
@@ -1,21 +1,41 @@
...
+        }
+
+        CheckExistingFile(overwrite, dropPath);
         using (var writer = File.CreateText(dropPath))
         {
             OutboxScriptBuilder.BuildDropScript(writer, sqlDialect);
         }
     }
+
+    static void CheckExistingFile(bool overwrite, string sagaFileName)
```
Comment (mikeminutillo) [inline src/ScriptBuilder/Writers/OutboxWriter.cs:34]: `sagaFileName` should be `outboxFileName`. Maybe this should just be `fileName`?

File: src/ScriptBuilder/Writers/OutboxWriter.cs [hunk 4]
```diff
@@ -1,21 +1,41 @@
...
 using NServiceBus.Persistence.Sql.ScriptBuilder;
 
 class OutboxWriter
 {
-    public static void WriteOutboxScript(string scriptPath, BuildSqlDialect sqlDialect)
+    public static void WriteOutboxScript(bool clean, bool overwrite, string scriptPath, BuildSqlDialect sqlDialect)
     {
         var createPath = Path.Combine(scriptPath, "Outbox_Create.sql");
-        File.Delete(createPath);
+        if (clean)
```
Comment (mikeminutillo) [inline src/ScriptBuilder/Writers/OutboxWriter.cs:10]: There's a lot of duplicated logic for each file and for each type of script we output. I wonder if it's worth pulling some of it into a base class and then having the outbox writer look something like this:

```csharp
class OutboxWriter : ScriptWriter
{
  ctor(clean, overwrite, scriptPath) : base(clean, overwrite, scriptPath) { ... }
  
  public void WriteOutboxScript(BuildSqlDialect sqlDialect)
  {
    WriteScript("Outbox_Create.sql", writer => OutboxScriptBuilder.BuildCreateScript(writer, sqlDialect));
    WriteScript("Outbox_Drop.sql", writer => OutboxScriptBuilder.BuildDropScript(writer, sqlDialect));
  }
}
```

File: src/ScriptBuilder/Writers/TimeoutWriter.cs [hunk 5]
```diff
@@ -1,21 +1,41 @@
...
+        }
+
+        CheckExistingFile(overwrite, dropPath);
         using (var writer = File.CreateText(dropPath))
         {
             TimeoutScriptBuilder.BuildDropScript(writer, sqlDialect);
         }
     }
+
+    static void CheckExistingFile(bool overwrite, string sagaFileName)
```
Comment (mikeminutillo) [inline src/ScriptBuilder/Writers/TimeoutWriter.cs:34]: Should be `timeoutFileName` or maybe just `fileName`?

File: src/ScriptBuilder/Writers/SubscriptionWriter.cs [hunk 6]
```diff
@@ -1,21 +1,41 @@
...
+        }
+
+        CheckExistingFile(overwrite, dropPath);
         using (var writer = File.CreateText(dropPath))
         {
             SubscriptionScriptBuilder.BuildDropScript(writer, sqlDialect);
         }
     }
+
+    static void CheckExistingFile(bool overwrite, string sagaFileName)
```
Comment (mikeminutillo) [inline src/ScriptBuilder/Writers/SubscriptionWriter.cs:34]: Should be `subscriptionFileName` or maybe just `fileName`.

File: src/ScriptBuilder/Saga/AllSagaDefinitionReader.cs [hunk 7]
```diff
@@ -27,12 +29,15 @@ public IEnumerable<SagaDefinition> GetSagas(Action<ErrorsException, TypeDefiniti
...
+                errors.Add(new Exception($"Error in '{type.FullName}' (Filename='{type.GetFileName()}'). Error: {exception.Message}", exception));
             }
         }
-        return sagas;
-    }
-
 
+        if (errors.Count > 0)
+        {
+            throw new AggregateException(errors);
```
Comment (mikeminutillo) [inline src/ScriptBuilder/Saga/AllSagaDefinitionReader.cs:38]: The way we used to do this, we would still generate a partial set of scripts if there were errors. i.e. If you had 5 sagas and one of them failed, you would get 4 scripts _before_ the error was thrown.

File: src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs [hunk 8]
```diff
@@ -5,4 +5,4 @@
     MySqlScripts = true,
     PostgreSqlScripts = true,
     OracleScripts = true,
-    ScriptPromotionPath = "$(SolutionDir)$(ProjectDir)Postfix")]
\ No newline at end of file
+    ScriptPromotionPath = @"$(SolutionDir)\Prod")]
```
Comment (mikeminutillo) [inline src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs:8]: Is this change still needed?

File: src/ScriptBuilderTask.Tests/InnerTaskTests.cs [hunk 9]
```diff
@@ -12,11 +11,11 @@ public void IntegrationTest()
     {
         var testDirectory = TestContext.CurrentContext.TestDirectory;
         var temp = Path.Combine(testDirectory, "InnerTaskTemp");
-        if (!Directory.Exists(temp))
+        if (Directory.Exists(temp))
         {
-            return;
```
Comment (mikeminutillo) [inline src/ScriptBuilderTask.Tests/InnerTaskTests.cs:17]: I'm confused by this change. Was this broken before? It sems like it was bailing out of the test early if the temp folder was missing, is that right?

File: src/ScriptBuilder/ScriptWriter.cs [hunk 10]
```diff
@@ -6,52 +6,103 @@
...
+            this.promotionFinder = promotionFinder;
+        }
+
+        public static void Write(string assemblyPath, string targetDirectory, Func<string, string> promotionPathFinder)
+        {
+            var writer = new ScriptWriter(assemblyPath, targetDirectory, promotionFinder: promotionPathFinder);
+            writer.Generate();
+        }
+
+        string GetScriptPath() => Path.Combine(destinationDirectory, "NServiceBus.Persistence.Sql");
```
[the remark this answers is not in the document]
Comment (HEskandari) [reply src/ScriptBuilder/ScriptWriter.cs:37]: done

File: src/ScriptBuilder/Writers/OutboxWriter.cs (hunk 4 above)
Comment (HEskandari) [reply src/ScriptBuilder/Writers/OutboxWriter.cs:10]: done

File: src/ScriptBuilder/Writers/TimeoutWriter.cs (hunk 5 above)
Comment (HEskandari) [reply src/ScriptBuilder/Writers/TimeoutWriter.cs:34]: won't be applicable after the refactoring to the base class.

File: src/ScriptBuilder/Writers/SubscriptionWriter.cs (hunk 6 above)
Comment (HEskandari) [reply src/ScriptBuilder/Writers/SubscriptionWriter.cs:34]: won't be applicable after the refactoring to the base class.

File: src/ScriptBuilder/Saga/AllSagaDefinitionReader.cs (hunk 7 above)
Comment (HEskandari) [reply src/ScriptBuilder/Saga/AllSagaDefinitionReader.cs:38]: I reverted back this error logging changes, so when used in MSBuild it will have the same old behavior.

File: src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs (hunk 8 above)
Comment (HEskandari) [reply src/ScriptBuilderTask.Tests.Target/AssemblyInfo.cs:8]: reverted back

File: src/ScriptBuilderTask.Tests/InnerTaskTests.cs (hunk 9 above)
Comment (HEskandari) [reply src/ScriptBuilderTask.Tests/InnerTaskTests.cs:17]: yes, this was incorrect. On the first, unless you create that directory manually the test would not generate anything.

File: src/ScriptBuilder/Writers/OutboxWriter.cs (hunk 3 above)
Comment (HEskandari) [reply src/ScriptBuilder/Writers/OutboxWriter.cs]: won't be applicable after the refactoring to the base class.

**Review by mikeminutillo — APPROVED** [review]
